framework/src/base/ComputeUserObjectsThread.C

```diff
@@ -70,6 +71,32 @@ ComputeUserObjectsThread::subdomainChanged()
     }
   }
 
+  // ShapeElementUserObject requested Jacobian variables
+  {
+    std::set<MooseVariable *> jacobian_moose_vars;
+
+    // Get the vectors of element user object pointers
+    const std::vector<ShapeElementUserObject *> global = _user_objects[_tid].shapeElementUserObjects(Moose::ANY_BLOCK_ID, _group);
+    const std::vector<ShapeElementUserObject *> block = _user_objects[_tid].shapeElementUserObjects(_subdomain, _group);
+
+    // Global ElementUserObjects
+    for (std::vector<ShapeElementUserObject *>::const_iterator it = global.begin(); it != global.end(); ++it)
+    {
+      const std::set<MooseVariable *> & mv_deps = (*it)->jacobianMooseVariables();
+      jacobian_moose_vars.insert(mv_deps.begin(), mv_deps.end());
+    }
+
+    // Block Restricted ElementUserObjects
+    for (std::vector<ShapeElementUserObject *>::const_iterator it = block.begin(); it != block.end(); ++it)
+    {
+      const std::set<MooseVariable *> & mv_deps = (*it)->jacobianMooseVariables();
+      jacobian_moose_vars.insert(mv_deps.begin(), mv_deps.end());
+    }
+
+    _jacobian_moose_vars.clear();
+    std::copy(jacobian_moose_vars.begin(), jacobian_moose_vars.end(), std::back_inserter(_jacobian_moose_vars));
```

Comment:
```
_jacobian_moose_vars.assign(jacobian_moose_vars.begin(), jacobian_moose_vars.end())
```
Would also work if you want a 1-liner.  It can either increase or decrease the size of the vector if necessary.